### ISSUE TITLE
Budget API Calls Returning Empty Spent Array

### ISSUE THREAD

**Issue by Unsantae** [firefly-iii/firefly-iii#9418]
### Support guidelines

- [x] I've read the [support guidelines](https://github.com/firefly-iii/firefly-iii/blob/main/.github/support.md)

### I've found a bug and checked that ...

- [x] ... [the documentation](https://docs.firefly-iii.org/) does not mention anything about my problem
- [x] ... there are no open or closed issues that are related to my problem
- [x] ... it's [definitely a Firefly III issue, not me](https://github.com/firefly-iii/firefly-iii/blob/main/.github/its_you_not_me.md)

### Description

API calls to the /api/v1/budgets endpoint return JSON with an empty "spent" array.

http://192.168.1.19:8080/api/v1/budgets?start=2024-10-01&end=2024-10-31

[
  {
    "data": [
      {
        "type": "budgets",
        "id": "2",
        "attributes": {
          "created_at": "2022-12-09T00:25:09-05:00",
          "updated_at": "2024-10-30T17:14:21-04:00",
          "active": false,
          "name": "Ayi",
          "order": 0,
          "notes": null,
          "auto_budget_type": "reset",
          "auto_budget_period": "monthly",
          "auto_budget_currency_id": "8",
          "auto_budget_currency_code": "USD",
          "auto_budget_amount": "300.00",
          "spent": []
        },
        "links": {
          "0": {
            "rel": "self",
            "uri": "/budgets/2"
          },
          "self": "http://172.19.0.22:8080/api/v1/budgets/2"
        }
      },
      {
        "type": "budgets",
        "id": "3",
        "attributes": {
          "created_at": "2022-12-09T00:25:33-05:00",
          "updated_at": "2023-03-05T16:31:20-05:00",
          "active": true,
          "name": "Groceries",
          "order": 1,
          "notes": null,
          "auto_budget_type": "reset",
          "auto_budget_period": "monthly",
          "auto_budget_currency_id": "8",
          "auto_budget_currency_code": "USD",
          "auto_budget_amount": "1200.00",
          "spent": []
        },
        "links": {
          "0": {
            "rel": "self",
            "uri": "/budgets/3"
          },
          "self": "http://172.19.0.22:8080/api/v1/budgets/3"
        }
      },
      {
        "type": "budgets",
        "id": "5",
        "attributes": {
          "created_at": "2022-12-09T00:26:31-05:00",
          "updated_at": "2024-07-01T07:39:36-04:00",
          "active": true,
          "name": "Children",
          "order": 2,
          "notes": null,
          "auto_budget_type": "reset",
          "auto_budget_period": "monthly",
          "auto_budget_currency_id": "8",
          "auto_budget_currency_code": "USD",
          "auto_budget_amount": "300.00",
          "spent": []
        },
        "links": {
          "0": {
            "rel": "self",
            "uri": "/budgets/5"
          },
          "self": "http://172.19.0.22:8080/api/v1/budgets/5"
        }
      },
      {
        "type": "budgets",
        "id": "6",
        "attributes": {
          "created_at": "2022-12-09T00:27:01-05:00",
          "updated_at": "2024-07-31T14:42:46-04:00",
          "active": true,
          "name": "Dining",
          "order": 3,
          "notes": null,
          "auto_budget_type": "reset",
          "auto_budget_period": "monthly",
          "auto_budget_currency_id": "8",
          "auto_budget_currency_code": "USD",
          "auto_budget_amount": "100.00",
          "spent": []
        },
        "links": {
          "0": {
            "rel": "self",
            "uri": "/budgets/6"
          },
          "self": "http://172.19.0.22:8080/api/v1/budgets/6"
        }
      }
    ],
    "meta": {
      "pagination": {
        "total": 4,
        "count": 4,
        "per_page": 200,
        "current_page": 1,
        "total_pages": 1
      }
    },
    "links": {
      "self": "http://1192.168.1.19:8080/api/v1/budgets?start=2024-11-01&end=2024-11-30&limit=200&page=1",
      "first": "http://192.168.1.19:8080/api/v1/budgets?start=2024-11-01&end=2024-11-30&limit=200&page=1",
      "last": "http://192.168.1.19:8080/api/v1/budgets?start=2024-11-01&end=2024-11-30&limit=200&page=1"
    }
  }
]

### Debug information

Debug information generated at 2024-10-31 20:21:19 for Firefly III version **v6.1.21**.

<table><thead><tr><th colspan="2">System information</th></tr><tr><th>Item</th><th>Value</th></tr></thead><tbody><tr><td>Firefly III</td><td>6.1.21 / v2.1.0 / 24 (exp. 24)</td></tr><tr><td>PHP version</td><td>8.3.12 (64bits) / apache2handler / Linux x86_64</td></tr><tr><td>BCscale</td><td>12</td></tr><tr><td>Error reporting</td><td>Display: Off, reporting: ALL errors</td></tr><tr><td>Max upload</td><td>67108864 (64 MB)</td></tr><tr><td>Database drivers</td><td>mysql, <strong>*pgsql*</strong>, sqlite, </td></tr><tr><td>Docker build</td><td><span>#</span>1131, base <span>#</span>91</td></tr></tbody></table><table><thead><tr><th colspan="2">Firefly III information</th></tr><tr><th>Item</th><th>Value</th></tr></thead><tbody><tr><td>Timezone</td><td>America/New_York + America/New_York</td></tr><tr><td>App environment</td><td>production, debug: false</td></tr><tr><td>Layout</td><td>v1</td></tr><tr><td>Logging</td><td>notice, stack / (empty)</td></tr><tr><td>Cache driver</td><td>redis</td></tr><tr><td>Default language and locale</td><td>en_US + equal</td></tr><tr><td>Trusted proxies</td><td>172.19.0.21</td></tr><tr><td>Login provider & user guard</td><td>eloquent / web</td></tr><tr><td>Login headers</td><td>N/A + N/A</td></tr><tr><td>Stateful domains</td><td></td></tr><tr><td>Last cron job</td><td>2024-10-31 04:00:00 (20 hours ago)</td></tr><tr><td>Mailer</td><td>smtp</td></tr></tbody></table><table><thead><tr><th colspan="2">User-specific information</th></tr><tr><th>Item</th><th>Value</th></tr></thead><tbody><tr><td>User</td><td><span>#</span>1 of 4</td></tr><tr><td>User flags</td><td><span title="Has rules">:wrench:</span> <span title="Has recurring transactions">:clock130:</span> <span title="Has subscriptions">:email:</span></td></tr><tr><td>Session start</td><td>2024-10-01 00:00:00</td></tr><tr><td>Session end</td><td>2024-10-31 23:59:59</td></tr><tr><td>View range</td><td>1M</td></tr><tr><td>User language</td><td>en_US</td></tr><tr><td>User locale</td><td>en_US</td></tr><tr><td>Locale(s) supported</td><td>en_US.utf8: :white_check_mark:<br>en_US.UTF-8: :white_check_mark:<br></td></tr><tr><td>User agent</td><td>Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/130.0.0.0 Safari/537.36</td></tr></tbody></table>

### Expected behaviour

I expect a JSON response in line with the API documentation that has a populated "spent" array for all active budgets with associated spending, per the below.

 "spent": [
          {
            "sum": "123.45",
            "currency_id": "5",
            "currency_code": "USD",
            "currency_symbol": "$",
            "currency_decimal_places": 2
          }

### Steps to reproduce

1. Send a api call to the desired endpoint http://192.168.1.19:8080/api/v1/budgets?start=2024-10-01&end=2024-10-31
2. Read the response which is missing the desired data.

**Comment by JC5** [comment]
Sorry, I can't replicate this. Submitting dates to this endpoint works for me ™ .

Are you sure there is data for the submitted period?

**Comment by Unsantae** [comment]
Yeah.  There is an entire month worth of transactions. This is a weird one
because it was working for about an hour while I was writing the program
that used the API, but then I took a break and when I came back it stopped
working.  I didn’t make any changes to the API call or the larger program.
I thought it might have something to do with time relative to the end of
the month, but I checked all of the time zones and that shouldn’t be an
issue.

This is the second time I’ve run into this issue.  The first time I worked
around it by calling this endpoint to get the budgetIDs and then making
individual calls to the budgetID endpoint for the spending data, so I was
actually a bit shocked when the normal budget endpoint worked correctly
initially.

-Nathan

**Comment by Unsantae** [comment]
I just tried again without making any changes and the endpoint returned
spending data for the month of November.  ¯\_(ツ)_/¯

I’ll keep messing with it, but you can delete the bug report if you want
since it does not seem to be a consistently reproducible issue.

-Nathan

**Comment by JC5** [comment]
I'll convert it to a discussion later today. I'm curious to see if we can find a root cause.
